### PR TITLE
Fix default handling of custom type during Go transformation

### PR DESCRIPTION
**Pull request by maxmcd**
For https://github.com/goadesign/goa/issues/3009

@raphael moving this here for easier discussion.

This change breaks the grpc test. The result expects something like this:
```go
if source.TimeSlice == "" {
    target.TimeSlice = []time.Time{}
}
```
Which is invalid, and I imagine grpc has no business supporting `time.Time`, but maybe I'm not thinking of valid use-cases for custom types within grpc? 

Would it be better to create a new test type to just test this kind of type with `go_transform`, or try and fix the grpc stuff, or something else...?

### CODE REVIEW

**Comment by raphael** [comment]
My inclination would be to try and make it work for gRPC as well since protobuf does support extensions via include which could mean new types that are no built-in into Goa. That being said `time.Time` wouldn't work as is because of the way github.com/golang/protobuf/ptypes/timestamp works (the code would have to generate a `AsTime()` call. So, if it ends up being too difficult to implement for gRPC it's probably OK to change the scope of the tests for now.

**Comment by maxmcd** [comment]
@raphael added code changes with attempt at supporting protobuf. 

Now just one weird issue. 

This test failure is rendering `string(source.TimeSlice)` for the conversion:
```
--- FAIL: TestProtoBufTransform (0.04s)
    --- FAIL: TestProtoBufTransform/to-protobuf-type (0.02s)
        --- FAIL: TestProtoBufTransform/to-protobuf-type/defaults-to-defaults (0.00s)
            protobuf_transform_test.go:147: invalid code, got:
                func transform() {
                	target := &WithDefaults{
                		Int:            int32(source.Int),
                		TimeSlice:      string(source.TimeSlice),
                		RequiredInt:    int32(source.RequiredInt),
```

The code that is rendering that here: https://github.com/maxmcd/goa/blob/32f3808786c79abb0a8cfb4d5738d3c89f60a3b8/grpc/codegen/protobuf_transform.go#L495-L502

Strangely (to me) `targetType` is equal to `""` once we get here, so it default back to the string type. I might be misunderstanding a little, but I think I would expect the `Meta` to be populated for the target `expr.AttributeExpr`, but it is `nil`. 

Any thoughts?

**Comment by maxmcd** [comment]
Ah, and should I replace `[]time.Time` with maybe `[]json.RawMessage` as a more representative test case? (Or maybe not important that it is a slice).

**Comment by maxmcd** [comment]
Ok, this somewhat invasive change fixes the generated code (and causes no other test failures), but not sure if this is taking things a little too far.

```diff
diff --git a/grpc/codegen/protobuf.go b/grpc/codegen/protobuf.go
index 4f852f77..f5994b73 100644
--- a/grpc/codegen/protobuf.go
+++ b/grpc/codegen/protobuf.go
@@ -213,7 +213,7 @@ func protoBufGoFullTypeName(att *expr.AttributeExpr, pkg string, s *codegen.Name
 	case expr.UserType, expr.CompositeExpr:
 		return protoBufFullMessageName(att, pkg, s)
 	case expr.Primitive:
-		return protoBufNativeGoTypeName(actual)
+		return protoBufNativeGoTypeName(att)
 	case *expr.Array:
 		return "[]" + protoBufGoFullTypeRef(actual.ElemType, pkg, s)
 	case *expr.Map:
@@ -377,8 +377,12 @@ func protoBufNativeMessageTypeName(t expr.DataType) string {
 // protoBufNativeGoTypeName returns the Go type corresponding to the given
 // primitive type generated by the protocol buffer compiler after compiling
 // the ".proto" file (in *.pb.go).
-func protoBufNativeGoTypeName(t expr.DataType) string {
-	switch t.Kind() {
+func protoBufNativeGoTypeName(att *expr.AttributeExpr) string {
+	typeName, _ := codegen.GetMetaType(att)
+	if typeName != "" {
+		return typeName
+	}
+	switch att.Type.Kind() {
 	case expr.BooleanKind:
 		return "bool"
 	case expr.IntKind:
@@ -402,7 +406,7 @@ func protoBufNativeGoTypeName(t expr.DataType) string {
 	case expr.BytesKind:
 		return "[]byte"
 	default:
-		panic(fmt.Sprintf("cannot compute native protocol buffer type for %T", t)) // bug
+		panic(fmt.Sprintf("cannot compute native protocol buffer type for %T", att.Type)) // bug
 	}
 }

diff --git a/grpc/codegen/protobuf_transform.go b/grpc/codegen/protobuf_transform.go
index 4d346bbb..4aad94e4 100644
--- a/grpc/codegen/protobuf_transform.go
+++ b/grpc/codegen/protobuf_transform.go
@@ -102,7 +102,9 @@ func protoBufTransform(source, target *expr.AttributeExpr, sourceVar, targetVar
 // generating protobuf types (as protogen itself won't honor these overrides).
 func removeMeta(att *expr.AttributeExpr) {
 	codegen.Walk(att, func(a *expr.AttributeExpr) error {
-		a.Meta = nil
+		delete(a.Meta, "struct:field:name")
+		delete(a.Meta, "struct:field:external")
+		// a.Meta = nil
 		return nil
 	})
 }
@@ -493,14 +495,14 @@ func convertType(source, target *expr.AttributeExpr, sourceVar string, ta *trans
 	if source.Type.Kind() != expr.IntKind && source.Type.Kind() != expr.UIntKind {
 		if sourceType != "" || targetType != "" {
 			if ta.proto || targetType == "" {
-				targetType = protoBufNativeGoTypeName(target.Type)
+				targetType = protoBufNativeGoTypeName(target)
 			}
 			return fmt.Sprintf("%s(%s)", targetType, sourceVar)
 		}
 		return sourceVar
 	}
 	if ta.proto {
-		return fmt.Sprintf("%s(%s)", protoBufNativeGoTypeName(target.Type), sourceVar)
+		return fmt.Sprintf("%s(%s)", protoBufNativeGoTypeName(target), sourceVar)
 	}
 	if targetType == "" {
 		targetType = codegen.GoNativeTypeName(target.Type)
```

**Comment by raphael** [comment]
The change above makes sense! It may allow for using protobuf extensions that add support for specific types so seems to be warranted. Thank you for the great contribution!

File: grpc/codegen/protobuf_transform_test.go [hunk 1]
```diff
@@ -258,11 +258,11 @@ const (
...
-		RequiredString: string(source.RequiredString),
-		DefaultBool:    bool(source.DefaultBool),
+		RequiredString: tdtypes.CustomString(source.RequiredString),
+		DefaultBool:    tdtypes.CustomBool(source.DefaultBool),
 	}
 	if source.Integer != nil {
-		target.Integer = int32(*source.Integer)
+		target.Integer = tdtypes.CustomInt(*source.Integer)
 	}
 }
```
Comment (maxmcd) [inline grpc/codegen/protobuf_transform_test.go:267]: I did end up having to change this test. I think this is a fix!? Not sure...

**Comment by maxmcd** [comment]
Nice! I think this is ready for review. `make` passes for me locally with the exception of one test:
```
--- FAIL: TestPkgImport (0.00s)
    --- FAIL: TestPkgImport/vendored-pkg (0.00s)
        convert_test.go:85: got goa.design/goa/vendor/github.com/some/pkg, expected github.com/some/pkg
FAIL
```

I think that's unrelated to my changes, and maybe (hopefully) is just due to some local environment difference.

**Comment by raphael** [comment]
Thank you!

**Pull request merged**